tensorboard/components/tf_backend/urlPathHelpers.ts

```diff
@@ -12,17 +12,42 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 ==============================================================================*/
-export function queryEncoder(params?: any): string {
-  // It's important that the keys be sorted, so we always grab the right file
-  // if we are talking to the backend generated by serialze_tensorboard.py
-  if (params == null) {
-    return '';
+/**
+ * A query parameter value can either be a string or a list of strings.
+ * A string `"foo"` is encoded as `key=foo`; a list `["foo", "bar"]` is
+ * encoded as `key=foo&key=bar`.
+ */
+export type QueryValue = string | string[];
+
+/**
+ * Add query parameters to a URL. Values will be URL-encoded. The URL
+ * may or may not already have query parameters. For convenience,
+ * parameters whose value is `undefined` will be dropped.
+ *
+ * For example, the following expressions are equivalent:
+ *
+ *     addParams("http://foo", {a: "1", b: ["2", "3+4"], c: "5"})
+ *     addParams("http://foo?a=1", {b: ["2", "3+4"], c: "5", d: undefined})
+ *     "http://foo?a=1&b=2&b=3%2B4&c=5"
+ */
+export function addParams(
+    baseURL: string,
+    params: {[param: string]: QueryValue}): string {
+  const keys = Object.keys(params).sort().filter(k => params[k] !== undefined);
+  if (!keys.length) {
+    return baseURL;  // no need to change '/foo' to '/foo?'
   }
-  const components = _.keys(params)
-                       .sort()
-                       .filter((k) => params[k] !== undefined)
-                       .map((k) => k + '=' + encodeURIComponent(params[k]));
-  const result = components.length ? '?' + components.join('&') : '';
-  // Replace parens for consistency with urllib.urlencode
-  return result.replace(/\(/g, '%28').replace(/\)/g, '%29');
+  const delimiter = baseURL.indexOf('?') !== -1 ? '&' : '?';
```

Comment:
One small issue that no one has bothered to fix (and maybe that's OK) is that the anchor section (`#...`) comes last in a URL
http://www.example.com:80/path/to/myfile.html?key1=value1&key2=value2#SomewhereInTheDocument
so if the base URL has an anchor, we want to make sure to append the GET params to the left of `#`.

However, TensorBoard logic never includes an anchor within a base URL. It wouldn't work because of course we sync the hash value with the current dashboard. I guess this is mainly an FYI for now.

Reply:
Correct; this does not take that into account. If necessary, we can easily update this function to do so. However, I doubt that it would be a useful feature; here is why. The `addParams` function is almost always going to be used to construct routes for AJAX requests, as opposed to actually setting `window.location`. But the server cannot see or use the fragment, so including a fragment would have no effect. It is only when directly navigating the user that the fragment would be useful. If this does indeed become necessary, we can add it in without breaking backward compatibility.